client/finality-grandpa/src/communication/mod.rs

```diff
@@ -665,14 +667,24 @@ impl<Block: BlockT> Sink<Message<Block>> for OutgoingMessages<Block>
 		}
 
 		// when locals exist, sign messages on import
-		if let Some((ref pair, _)) = self.locals {
+		if let Some(ref public) = self.locals {
+			let keystore = match &self.keystore {
+				Some(keystore) => keystore.clone(),
+				None => {
+					return Err(Error::Signing("Cannot sign without a keystore".to_string()))
+				}
+			};
+
 			let target_hash = *(msg.target().0);
 			let signed = sp_finality_grandpa::sign_message(
+				keystore.clone(),
```

Comment:
seems like a lot of clones are happening on keystore. makes me think that problems are being solved by slapping on `.clone()` instead of reasoning about ownership

Reply:
The keystore is defined as
```
	keystore: Option<BareCryptoStorePtr>,
```
in `OutgoingMessage` which is `Arc<parking_lot::RWLock<dyn BareCryptoStore>>` which is fine to clone as you certainly know. And since i had to slap it inside `OutgoingMessage` so that it can be used when polling the sink.

Reply:
But keystore is already `clone()`d a few lines above. It doesn't seem necessary to clone it again.

Reply:
I see your point... fixing.